### ISSUE TITLE
Small question about the code

### ISSUE THREAD

**Issue by LifeBeyondExpectations** [leoshine/Spherical_Regression#4]
First of all, thank you for sharing the nice paper. I am very excited to encounter such a fancy idea.

I have a question for the code.
For my understanding, the function "exp_Normalization" is the one that you mentioned as "Regression with S_exp" in the paper.
https://github.com/leoshine/Spherical_Regression/blob/5c9c352e6d47844117dbc6ed92753ef1ad60e8c6/pylibs/pytorch_util/torch_3rd_funcs.py#L108

But this function is not used, instead, I check the code as below,
https://github.com/leoshine/Spherical_Regression/blob/ba4a41dc04f4b8fdb80169165c3254c541a7bc4a/S3.3D_Rotation/regQuatNet/regQuatNet.py#L203

It looks weird.
As far as my understanding, raw latent embedding O undergoes:

(1) softmax 
https://github.com/leoshine/Spherical_Regression/blob/ba4a41dc04f4b8fdb80169165c3254c541a7bc4a/S3.3D_Rotation/regQuatNet/regQuatNet.py#L250

(2) square root
https://github.com/leoshine/Spherical_Regression/blob/ba4a41dc04f4b8fdb80169165c3254c541a7bc4a/S3.3D_Rotation/regQuatNet/regQuatNet.py#L255

which is not the same eq. (9) in the paper.
However, the function "exp_Normalization" in the "pylibs/pytorch_util/torch_3rd_funcs.py', I am fully understood and agree with the code.

In short, my questions are:
1. Is there any reason that you do not use that function intentionally??
2. Sorry for my excuses, is there anything that I missed ??..

**Comment by LifeBeyondExpectations** [comment]
Sorry, my bad. I finally understand. Just scale problem.

One last question is that you detach the norm in your code.
https://github.com/leoshine/Spherical_Regression/blob/5c9c352e6d47844117dbc6ed92753ef1ad60e8c6/pylibs/pytorch_util/torch_3rd_funcs.py#L140
just want to know, what if I do not detach the norm ??
For my results, the network is not trained properly. 

1. Do you have any reason for this coding design? 
2. Is it related to the fact that you do not use the function "exp_Normalization" ??

**Comment by leoshine** [comment]
Thanks for your interests. 

First of all, some functions in pytorch_util/torch_3rd_funcs.py are indeed not used. They were used for verification purpose (of the exact Sexp formula in the paper).

Second, indeed the implementation in S3.3D_Rotation/regQuatNet/regQuatNet.py looks different from the paper, but they are actually the same. The reason for such an implementation trick to increase numeric stability. This is similar to the implementation of softmax+cross_entropy loss used in a standard classification problem where the network instead directly predict the raw pre-softmax output O (a.k.a the logits), it predicts log(O).

Last:
 1. not detaching the norm makes the computational graph complex, which does not comply with the intention of normalization operation. In other words, if the norm does not detach from the computational graph and thus receive gradient, the network becomes sensitive the magnitude (norm) of the raw output signal. However, I didn't test this and don't know the actual behavior of training when the norm is not detached.
 2. explained as above.
